File: sdk/python/meetingbot/webhooks.py
"""Webhook signature verification helpers.

JustHereToListen.io signs every webhook delivery with HMAC-SHA256 over
``f"{timestamp}.{body}"``. The signature is sent in ``X-MeetingBot-Signature``
(``sha256=<hex>``) and the timestamp in ``X-MeetingBot-Timestamp`` (Unix seconds).

Receivers MUST:
1. Verify the signature with a constant-time compare.
2. Reject deliveries whose timestamp is older than ``max_age_seconds`` to
   prevent replay (default: 300 s).
"""

from __future__ import annotations

import hashlib
import hmac
import time

from .exceptions import MeetingBotError


class WebhookVerificationError(MeetingBotError):
    """Raised when a webhook signature or timestamp fails verification."""


def verify_webhook(
    body: str | bytes,
    timestamp: str | int,
    signature: str,
    secret: str,
    *,
    max_age_seconds: int = 300,
    now: float | None = None,
) -> None:
    """Verify a webhook delivery's signature and freshness.

    :param body: Raw request body (str or bytes), exactly as received.
    :param timestamp: ``X-MeetingBot-Timestamp`` header value (Unix seconds).
    :param signature: ``X-MeetingBot-Signature`` header value (``sha256=<hex>``).
    :param secret: Your webhook's signing secret.
    :param max_age_seconds: Maximum acceptable age in seconds (default: 300).
    :param now: Override the current time (for testing).

    :raises WebhookVerificationError: if the signature is invalid or the
        timestamp is missing, malformed, or outside the freshness window.
    """
    if not signature or not signature.startswith("sha256="):
        raise WebhookVerificationError("Missing or malformed signature header")
    try:
        ts_int = int(str(timestamp).strip())
    except (TypeError, ValueError) as exc:
        raise WebhookVerificationError(f"Invalid timestamp: {timestamp!r}") from exc

    current = time.time() if now is None else now
    if abs(current - ts_int) > max_age_seconds:
        raise WebhookVerificationError(
            f"Timestamp {ts_int} is outside the {max_age_seconds}s freshness window"
        )

    body_bytes = body.encode() if isinstance(body, str) else body
    signed_payload = f"{ts_int}.".encode() + body_bytes
    expected = "sha256=" + hmac.new(secret.encode(), signed_payload, hashlib.sha256).hexdigest()
    if not hmac.compare_digest(expected, signature):
        raise WebhookVerificationError("Signature mismatch")

File: sdk/js/src/webhooks.ts
/**
 * Webhook signature verification.
 *
 * JustHereToListen.io signs every webhook delivery with HMAC-SHA256 over
 * `${timestamp}.${body}`. The signature is sent in `X-MeetingBot-Signature`
 * (`sha256=<hex>`) and the timestamp in `X-MeetingBot-Timestamp` (Unix seconds).
 *
 * Receivers MUST:
 * 1. Verify the signature with a constant-time compare.
 * 2. Reject deliveries whose timestamp is older than `maxAgeSeconds` to
 *    prevent replay (default: 300 s).
 */

import { createHmac, timingSafeEqual } from "node:crypto";

import { MeetingBotError } from "./errors.js";

export class WebhookVerificationError extends MeetingBotError {
  constructor(message: string) {
    super(message);
    this.name = "WebhookVerificationError";
  }
}

export interface VerifyWebhookOptions {
  /** Maximum acceptable age in seconds (default: 300). */
  maxAgeSeconds?: number;
  /** Override the current time in seconds, for testing. */
  now?: number;
}

/**
 * Verify a webhook delivery's signature and freshness.
 *
 * Throws {@link WebhookVerificationError} if the signature is invalid or the
 * timestamp is missing, malformed, or outside the freshness window.
 */
export function verifyWebhook(
  body: string | Buffer,
  timestamp: string | number,
  signature: string,
  secret: string,
  options: VerifyWebhookOptions = {},
): void {
  const { maxAgeSeconds = 300, now } = options;

  if (!signature || !signature.startsWith("sha256=")) {
    throw new WebhookVerificationError("Missing or malformed signature header");
  }

  const tsInt = Number.parseInt(String(timestamp).trim(), 10);
  if (!Number.isFinite(tsInt)) {
    throw new WebhookVerificationError(`Invalid timestamp: ${String(timestamp)}`);
  }

  const current = now ?? Math.floor(Date.now() / 1000);
  if (Math.abs(current - tsInt) > maxAgeSeconds) {
    throw new WebhookVerificationError(
      `Timestamp ${tsInt} is outside the ${maxAgeSeconds}s freshness window`,
    );
  }

  const bodyBuf = typeof body === "string" ? Buffer.from(body, "utf8") : body;
  const signedPayload = Buffer.concat([Buffer.from(`${tsInt}.`, "utf8"), bodyBuf]);
  const expected =
    "sha256=" + createHmac("sha256", secret).update(signedPayload).digest("hex");

  const a = Buffer.from(expected, "utf8");
  const b = Buffer.from(signature, "utf8");
  if (a.length !== b.length || !timingSafeEqual(a, b)) {
    throw new WebhookVerificationError("Signature mismatch");
  }
}
